File: Algorithms/Sorting/Quicksort 1 - Partition/solution.js
// https://www.hackerrank.com/challenges/quicksort1/problem
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.replace(/\s*$/, '')
        .split('\n')
        .map(str => str.replace(/\s*$/, ''));

    main();
});

function readLine() {
    return inputString[currentLine++];
}

function quickSort(arr) {
    var pivot = arr[0];
    var equal = [], left = [], right = [];

    for(var i = 0; i < arr.length; i++){
        if(arr[i] < pivot)
            left.push(arr[i]);
        else if(arr[i] == pivot)
            equal.push(arr[i]);
        else
            right.push(arr[i]);
    }

    return left.concat(equal).concat(right);
}

function main() {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    const n = parseInt(readLine(), 10);

    const arr = readLine().split(' ').map(arrTemp => parseInt(arrTemp, 10));

    let result = quickSort(arr);

    ws.write(result.join(" ") + "\n");

    ws.end();
}

File: Algorithms/Sorting/Quicksort 1 - Partition/solution.cs
// https://www.hackerrank.com/challenges/quicksort1/problem
using System.CodeDom.Compiler;
using System.Collections.Generic;
using System.Collections;
using System.ComponentModel;
using System.Diagnostics.CodeAnalysis;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.Serialization;
using System.Text.RegularExpressions;
using System.Text;
using System;

class Solution {

    static int[] quickSort(int[] arr) {
        int pivot = arr[0];
        List<int> equal = new List<int>();
        List<int> left = new List<int>();
        List<int> right = new List<int>();

        for(int i = 0; i < arr.Length; i++){
            if(arr[i] < pivot)
                left.Add(arr[i]);
            else if(arr[i] == pivot)
                equal.Add(arr[i]);
            else
                right.Add(arr[i]);
        }

        List<int> combined = new List<int>(left);

        combined.AddRange(equal);
        combined.AddRange(right);

        return combined.ToArray();
    }

    static void Main(string[] args) {
        TextWriter textWriter = new StreamWriter(@System.Environment.GetEnvironmentVariable("OUTPUT_PATH"), true);

        int n = Convert.ToInt32(Console.ReadLine());

        int[] arr = Array.ConvertAll(Console.ReadLine().Split(' '), arrTemp => Convert.ToInt32(arrTemp))
        ;
        int[] result = quickSort(arr);

        textWriter.WriteLine(string.Join(" ", result));

        textWriter.Flush();
        textWriter.Close();
    }
}